pkg/daemon/daemon.go

```diff
@@ -144,6 +144,10 @@ func (d *Daemon) State() (types.DaemonState, error) {
 		return types.DaemonStateUnknown, err
 	}
 
+	// Protect daemon client when it's being reset.
+	d.Lock()
```

Comment:
How can we ensure no new race is brought for new operations? In other word, can we simply the `ensureClient(), Lock(), Unlock()` code?

Reply:
True. We need to figure out a code restriction framework to close the race once and for all. But now let's fix the race problem first since it occasionally fails e2e test